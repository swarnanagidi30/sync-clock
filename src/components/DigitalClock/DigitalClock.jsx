import React, { useState, useEffect, useRef, useCallback } from 'react'
import './DigitalClock.scss';
import ClockEditor from '../ClockEditor';

export default function DigitalClock(props) {
    const { date, datediff, onDateChange } = props;
    const refClockInterval = useRef(null);

    const [hours, setHours] = useState(date.getHours());
    const [minutes, setMinutes] = useState(date.getMinutes());
    const [seconds, setSeconds] = useState(date.getSeconds());
    const [showEditor, setShowEditor] = useState(false)

    const handleDate = useCallback(
        () => {
            const formatTime = (time) => {
                return time < 10 ? `0${time}` : time;
            }
            date.setHours(date.getHours() + (datediff || 0));
            setHours(formatTime(date.getHours()));
            setMinutes(formatTime(date.getMinutes()));
            setSeconds(formatTime(date.getSeconds()));
        },
        [datediff, date],
    );

    useEffect(() => {

        refClockInterval.current = setInterval(handleDate, 1000);
        return () => {
            clearInterval(refClockInterval.current);
            console.log('cleared');
        }
    }, [datediff, refClockInterval, date, handleDate])

    const handleEdit = () => {
        clearInterval(refClockInterval.current);
        setShowEditor(true);
    }
    const handleEditClose = (date) => {
        refClockInterval.current = setInterval(handleDate, 1000);
        setShowEditor(false);
        onDateChange(date);
    }
    return (
        <div className="digital-root">
            <div className="digital-clock">
                <div className="inner">
                    <span className="span" > {hours}</span>
                :<span className="span">{minutes}</span>
                :<span className="span">{seconds}</span>
                </div>
                <div className="clock-setting">
                    <i className="fa fa-cog" aria-hidden="true" onClick={handleEdit} ></i>
                </div>
            </div>
            {showEditor && <ClockEditor date={date} onClose={handleEditClose} />}
        </div>
    )
}
