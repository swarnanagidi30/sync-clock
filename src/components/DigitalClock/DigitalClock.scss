/* Google font */
@import url('https://fonts.googleapis.com/css?family=Orbitron');

$clockSize: 360px;
$clockPadding: $clockSize / 14.4;
$clockBackground: #F7F7F7;

.digital-root{
    position: relative;
    .modal-content{
        height: 70%;
        width: 50%;
    }
.digital-clock {
    width: 550px;
    font-family: 'Orbitron', sans-serif;
    font-size: 90px;
    margin: 20px;
    margin-top: 110px;
    padding: 20px;
    background-color: $clockBackground;
    color: #17D4FE;
    letter-spacing: 7px;
    border-radius: 40px;
    box-shadow: 0 18px 40px rgba(0, 0, 0, 0.15);
    position: relative;
    
}
&:hover{
    .clock-setting{
        display: block;
    }
}

.clock-setting{
    display: none;
    font-size: 17px;
    color: #000;
    position: absolute;
    right: -12px;
    top: -5px;
    // &:hover{
    //     background-color: white;

    // }
}
}