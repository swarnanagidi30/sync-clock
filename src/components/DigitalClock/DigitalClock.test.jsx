import React from 'react';
import { render } from '@testing-library/react';
import DigitalClock from './DigitalClock';
import renderer from 'react-test-renderer';

it('renders correctly', () => {
    const date = new Date(2020, 11, 20, 10, 50, 55);
    const tree = renderer
        .create(<DigitalClock date={date} />)
        .toJSON();
    expect(tree).toMatchSnapshot();
});

test('renders Digital Clock time', () => {
    const date = new Date(2020, 11, 20, 10, 50, 55);
    const { getByText } = render(<DigitalClock date={date} />);
    const hourElement = getByText(/10/i);
    expect(hourElement).toBeInTheDocument();

    const minutesElement = getByText(/50/i);
    expect(minutesElement).toBeInTheDocument();

    const secondsElement = getByText(/55/i);
    expect(secondsElement).toBeInTheDocument();
});
