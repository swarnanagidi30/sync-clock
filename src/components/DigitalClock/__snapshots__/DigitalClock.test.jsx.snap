// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="digital-root"
>
  <div
    className="digital-clock"
  >
    <div
      className="inner"
    >
      <span
        className="span"
      >
         
        10
      </span>
      :
      <span
        className="span"
      >
        50
      </span>
      :
      <span
        className="span"
      >
        55
      </span>
    </div>
    <div
      className="clock-setting"
    >
      <i
        aria-hidden="true"
        className="fa fa-cog"
        onClick={[Function]}
      />
    </div>
  </div>
</div>
`;
